### PR TITLE
Remove configuration options for direct TCP replication.

### PR DESCRIPTION
**Pull request by clokep**
Removes the ability to configure legacy direct TCP replication. Workers now require Redis to run.

This was deprecated in v1.18.0 from 2 years ago: 2020-07-30 (see [release notes](https://github.com/matrix-org/synapse/blob/master/docs/changelogs/CHANGES-2020.md#synapse-1180-2020-07-30), [upgrade notes](https://github.com/matrix-org/synapse/blob/develop/docs/upgrade.md#redis-replication-is-now-recommended-in-lieu-of-tcp-replication)). The [blog post](https://matrix.org/blog/2020/07/30/synapse-1-18-0-released) unfortunately doesn't seem to mention this.

Part of #11728. Requires matrix-org/sytest#1282.

### CODE REVIEW

File: synapse/config/server.py [hunk 1]
```diff
@@ -165,7 +165,6 @@ def generate_ip_set(
     "http",
     "metrics",
     "manhole",
-    "replication",
```
Comment (clokep) [inline synapse/config/server.py:168]: We talked briefly about whether providing `replication` should show a deprecation warning instead of error. This has been deprecated for almost two years, so seems reasonable to just remove.

File: changelog.d/13647.removal [hunk 2]
```diff
@@ -0,0 +1 @@
+Remove the ability to direct TCP replication with workers. Workers now require using Redis.
```
Comment (reivilibre) [inline changelog.d/13647.removal:1]:
```suggestion
Remove the ability to use direct TCP replication with workers. Direct TCP replication was deprecated in Synapse v1.18.0. Workers now require using Redis.
```
I personally think this context is useful enough to have directly in the changelog.

File: docs/upgrade.md [hunk 3]
```diff
@@ -89,6 +89,22 @@ process, for example:
...
+
+## Direct TCP replication is no longer supported: migrate to Redis
+
+Redis support was added in v1.13.0 with it becoming the recommended method in
+v1.18.0. It replaced the old direct TCP connections (which was deprecated as of
+v1.18.0) to the main process. With Redis, rather than all the workers connecting
+to the main process, all the workers and the main process connect to Redis,
+which relays replication commands between processes. This can give a significant
+cpu saving on the main process and will be a prerequisite for upcoming
+performance improvements.
```
Comment (reivilibre) [inline docs/upgrade.md:102]:
```suggestion
CPU saving on the main process and is a prerequisite for upcoming
performance improvements.
```

File: docs/upgrade.md [hunk 4]
```diff
@@ -89,6 +89,22 @@ process, for example:
...
+
+To migrate to Redis add the [`redis` config](./workers.md#shared-configuration),
+and remove the TCP `replication` listener from master and `worker_replication_port`
```
Comment (reivilibre) [inline docs/upgrade.md:105]: I don't know what the TCP configuration looked like, but I think that you still need the `replication` listener in the `listeners` block of the config, in order to serve the portions of replication that happen over HTTP requests?
This paragraph risks misleading me into removing the HTTP listener. Think it would be worth clarifying if you know how?

Edit: I think I see. You still want the `replication` *resource* on the `type: http` listener, but there was previously a `type: replication` listener.

File: docs/workers.md [hunk 5]
```diff
@@ -32,13 +32,8 @@ stream between all configured Synapse processes. Additionally, processes may
...
 for a reply ─ such as sending an event.
 
-Redis support was added in v1.13.0 with it becoming the recommended method in
-v1.18.0. It replaced the old direct TCP connections (which is deprecated as of
-v1.18.0) to the main process. With Redis, rather than all the workers connecting
-to the main process, all the workers and the main process connect to Redis,
-which relays replication commands between processes. This can give a significant
-cpu saving on the main process and will be a prerequisite for upcoming
-performance improvements.
+All the workers and tha main process connect to Redis, which relays replication
```
Comment (reivilibre) [inline docs/workers.md:35]:
```suggestion
All the workers and the main process connect to Redis, which relays replication
```

File: synapse/config/server.py (hunk 1 above)
Comment (reivilibre) [reply synapse/config/server.py:168]: Agreed

File: .github/workflows/tests.yml [hunk 6]
```diff
@@ -236,7 +235,6 @@ jobs:
           - sytest-tag: buster
             postgres: postgres
             workers: workers
-            redis: redis
```
Comment (reivilibre) [inline .github/workflows/tests.yml:239]: Noting that this now is only different from the previous run in the list by using `postgres` instead of `multi-postgres`. Happy to accept that this may still be worth testing though.

File: synapse/app/homeserver.py [hunk 7]
```diff
@@ -290,16 +289,6 @@ def start_listening(self) -> None:
                     manhole_settings=self.config.server.manhole_settings,
                     manhole_globals={"hs": self},
                 )
-            elif listener.type == "replication":
-                services = listen_tcp(
-                    listener.bind_addresses,
-                    listener.port,
-                    ReplicationStreamProtocolFactory(self),
```
Comment (reivilibre) [inline synapse/app/homeserver.py:297]: noting that this is still used in tests, so `ReplicationStreamProtocolFactory` isn't unreachable yet, hence reasonable to not remove it absolutely immediately in this PR

**Review by reivilibre — APPROVED** [review]
This looks good to me; minor remarks follow.

Wonder if we would want to sneak in an announcement into the changelog for the release that's about to go out tomorrowish, just as a final warning? I don't think we **need** to, since it's been deprecated for such a long time, but also it's never been explicitly stated when it's going.
Could also consider forewarning in the announcements room?

File: docs/upgrade.md (hunk 4 above)
Comment (clokep) [reply docs/upgrade.md:105]: Thanks, @turt2live had the same feedback and I failed to mark that down here. 👍

File: .github/workflows/tests.yml (hunk 6 above)
Comment (clokep) [reply .github/workflows/tests.yml:239]: Correct! Maybe that is not worth testing though, would be worth hearing other's thoughts.

File: docs/upgrade.md (hunk 3 above)
Comment (clokep) [reply docs/upgrade.md:102]: Thanks! I copied this from the docs and didn't proof it too hard. 👍

**Comment by clokep** [comment]
> Wonder if we would want to sneak in an announcement into the changelog for the release that's about to go out tomorrowish, just as a final warning? I don't think we **need** to, since it's been deprecated for such a long time, but also it's never been explicitly stated _when_ it's going.

It might make sense to just put it at the top of the changelog and in the blog post?

File: docs/upgrade.md (hunk 4 above)
Comment (clokep) [reply docs/upgrade.md]: I updated this a bit in 398c493d68f71a5b4a75c3839ea559aeeebd51ae, do you think it is clearer now?

**Review by richvdh — COMMENTED** [review]
can we make `worker_replication_port` make an error?

File: .github/workflows/tests.yml (hunk 6 above)
Comment (clokep) [reply .github/workflows/tests.yml:239]: According to @erikjohnston:

> Ah, OK. I think that makes `(buster, postgres, workers, redis)` pretty redundant, compared with `(buster, multi-postgres, workers)`

**Comment by clokep** [comment]
> can we make `worker_replication_port` make an error?

I've now done that. Since I was adding an error message I also added one for adding a replication-type listener.

File: synapse/config/server.py (hunk 1 above)
Comment (clokep) [reply synapse/config/server.py:168]: 76dcd63464564952c2ac25bd30ca6f87e03d8fbd adds an error message.

File: tests/utils.py [hunk 8]
```diff
@@ -135,7 +135,6 @@ def default_config(
         "enable_registration_captcha": False,
         "macaroon_secret_key": "not even a little secret",
         "password_providers": [],
-        "worker_replication_url": "",
```
Comment (clokep) [inline tests/utils.py:138]: This is not really related, but it was removed in 2017 in #2117.

File: docs/upgrade.md (hunk 4 above)
Comment (reivilibre) [reply docs/upgrade.md]: Yep, that's good — thanks!

**Review by reivilibre — APPROVED** [review]
Good stuff! Thanks for starting us off; this should make it much easier to start pulling all of it out. Been wanting to do that for a while :)
